lib/ansible/modules/web_infrastructure/ansible_tower/tower_group.py

```diff
@@ -162,14 +204,17 @@ def main():
             if state == 'present':
                 result = group.modify(**params)
                 json_output['id'] = result['id']
+                json_output['changed'] = result['changed']
+                for p_group in parent_groups:
+                    res = group.associate(group=result['id'], parent=p_group, inventory=params['inventory'])
```

Comment:
Something isn't right here. On line 177-178 you did `group.get(name=p_group, inventory=params['inventory'])`, implying that entries in the `parent_groups` list are names. The usage here `parent=p_group` implies that those same entries are ids.

Just like you did `result['id']`, I think you should extract the id from `group.get(name=p_group, inventory=params['inventory'])`, so really just save those values from the validation step, and use those values here instead of the unvalidated `parent_groups` parameter.

Reply:
I think I actually did it that way in tower_host. the tower_cli host object didn't accept the names. But the group object does so I didn't make the same change here. 

The names are validated because the module would return a failure from line 180 before trying to use them, but I do see that its less obvious it works like this

If you think it would be better to use the IDs here I will make that change.

Reply:
Yes, better to use IDs. Even though the group associate command accepts names, it will be doing its own extra lookup. Even worse, it will be following a different uniqueness rule than what you've done here. You looked up groups within the same inventory as the host (which is correct), and tower-cli has a very significant limitation where it will look it up by name only.

I want to have a broader conversation about the direction the CLI and the modules will go in, and was just discussing with @rooftopcellist. Resolving the input arguments going into the module to produce lookup keyword args for the server is a very nuanced and non-trivial task. On the CLI side, we can't do that without really changing the framework used, so I'm open to allowing the modules to do more of the work on their side, which is pretty much what you're already doing here.

Reply:
Ok. I'll change it to lookup the IDs in the validation and use those in the associate call.